### PR TITLE
DATA-44 Adding API endpoint to get the activity completion data per user.

### PR DESCRIPTION
**Pull request by Squirrel18**
## Description:

This PR adds a new API endpoint to get the activity completion data per user instead of per course.

## Reviewers: 
- [ ] @andrey-canon 
- [ ] @diegomillan

### CODE REVIEW

File: openedx_proversity_reports/api/v0/views.py [hunk 1]
```diff
@@ -242,3 +247,79 @@ def post(self, request):
...
+                "passing_score": 0.75,
+                "users": [
+                    "honor@example.com",
+                    "audit@example.com"
+                ]
+            }
+        **Example Requests**:
+            POST /proversity-reports/proversity-reports/api/v0/user-activity-completion-data
+        **Response Values**:
+            * result: Activity completion data per user.
```
Comment (andrey-canon) [inline openedx_proversity_reports/api/v0/views.py]: please write an example

File: openedx_proversity_reports/serializers.py [hunk 2]
```diff
@@ -55,3 +58,35 @@ class ActivityCompletionReportSerializer(serializers.Serializer):
...
+            obj: Serializer fields object.
+        Returns:
+            course_keys: List containing opaque_keys.edx.keys.CourseKey course instances.
+        """
+        course_keys = []
+
+        for course_id in obj.get('course_ids', []):
+            try:
+                course_key = CourseKey.from_string(course_id)
+                course_keys.append(course_key)
```
Comment (andrey-canon) [inline openedx_proversity_reports/serializers.py]: merge with the previous line

File: openedx_proversity_reports/utils.py [hunk 3]
```diff
@@ -287,3 +287,78 @@ def get_attribute_from_module(module, attribute_name):
...
+        user_email_list: List containing the emails of the users.
+    Returns:
+        exisiting_user_list: List containing django.contrib.auth.models.User instances.
+    """
+    exisiting_user_list = []
+
+    for user_email in user_email_list:
+        try:
+            user = User.objects.get(email=user_email)
+            exisiting_user_list.append(user)
```
Comment (andrey-canon) [inline openedx_proversity_reports/utils.py]: merge with the previous line

File: openedx_proversity_reports/api/v0/views.py [hunk 4]
```diff
@@ -242,3 +247,79 @@ def post(self, request):
...
+    permission_classes = (permissions.IsAuthenticated, get_staff_or_owner())
+
+    def post(self, request):
+        """
+        Return a JSON representation of the activity completion data by user.
+
+        **Params**
+            users: List of the email of the users. This parameter must contain at least one value.
+            course_ids: List of course ids to filter only by those courses. This parameter can be empty.
+            block_types: List of the block types to get the completion data. This parameter must contain at least one value.
```
Comment (andrey-canon) [inline openedx_proversity_reports/api/v0/views.py]: same comment as above

File: openedx_proversity_reports/utils.py [hunk 5]
```diff
@@ -287,3 +287,78 @@ def get_attribute_from_module(module, attribute_name):
...
+
+    Args:
+        user: django.contrib.auth.models.User instance.
+    Returns:
+        List containing opaque_keys.edx.keys.CourseKey course instances.
+    """
+    course_enrollment_objects = get_course_enrollment().objects
+    user_course_enrollments = course_enrollment_objects.filter(
+        user__id=user.id,
+        user__courseaccessrole__id=None,
```
Comment (andrey-canon) [inline openedx_proversity_reports/utils.py]: why is this necessary ?

File: openedx_proversity_reports/utils.py [hunk 6]
```diff
@@ -287,3 +287,78 @@ def get_attribute_from_module(module, attribute_name):
...
+        }
+    """
+    required_activities_data = {}
+    total_activities = user_data.get('total_activities', 0)
+
+    if total_activities:
+        total = int(total_activities)
+        # Create as many 'required_activity_' as the total number of activities.
+        for activity_number in range(1, total + 1): # Plus 1, because the stop argument it's not inclusive.
+            required_activity_number = user_data.get('required_activity_{}'.format(activity_number), '')
```
Comment (andrey-canon) [inline openedx_proversity_reports/utils.py]: if this is `completed` or `not_completed`, this should be call different, maybe `required_activity_status` and the default value should be `not_completed`

File: openedx_proversity_reports/api/v0/views.py (hunk 1 above)
Comment (Squirrel18) [reply openedx_proversity_reports/api/v0/views.py]: Yes, I forget it.

File: openedx_proversity_reports/reports/activity_completion_report.py [hunk 7]
```diff
@@ -29,12 +30,28 @@ class GenerateCompletionReport(object):
...
         self.users = users
         self.course_key = course_key
-        self.course = get_course_by_id(self.course_key)
         self.required_block_types = block_types
         self.required_block_ids = self.get_course_required_block_ids(required_block_ids)
         self.passing_score = passing_score
 
 
+    @classmethod
+    def activity_completion_data_per_user(cls, user, course_key, block_types, passing_score):
```
[the remark this answers is not in the document]
Comment (Squirrel18) [reply openedx_proversity_reports/reports/activity_completion_report.py]: It is better, since i need to overload the method and explain why it exists.

File: openedx_proversity_reports/reports/activity_completion_report.py [hunk 8]
```diff
@@ -174,6 +191,29 @@ def get_course_required_block_ids(self, required_block_ids):
...
+        Returns:
+            user_activity_completion_data: Dict containing the activity completion data.
+        """
+        activity_completion_data = self.generate_report_data()
+
+        if not activity_completion_data:
+            return {}
+
+        user_activity_completion_data = {
+            'total_activities': activity_completion_data[0].get('total_activities', ''),
```
[the remark this answers is not in the document]
Comment (Squirrel18) [reply openedx_proversity_reports/reports/activity_completion_report.py]: Addresed.

File: openedx_proversity_reports/reports/activity_completion_report.py [hunk 9]
```diff
@@ -174,6 +191,29 @@ def get_course_required_block_ids(self, required_block_ids):
...
+            'course_is_complete': activity_completion_data[0].get('course_is_complete', False),
+            'completed_activities': activity_completion_data[0].get('completed_activities', ''),
```
[the remark this answers is not in the document]
Comment (Squirrel18) [reply openedx_proversity_reports/reports/activity_completion_report.py]: Addresed.

File: openedx_proversity_reports/api/v0/views.py [hunk 10]
```diff
@@ -242,3 +247,79 @@ def post(self, request):
...
+            for course_id in user_course_ids:
+                activity_completion_data = GenerateCompletionReport.activity_completion_data_per_user(
+                    user=user,
+                    course_key=course_id,
+                    block_types=serialized_data.data.get('block_types', []),
+                    passing_score=serialized_data.data.get('passing_score', 0),
+                ).get_user_completion_data()
+
+                course_data = {
+                    str(course_id): activity_completion_data,
```
[the remark this answers is not in the document]
Comment (Squirrel18) [reply openedx_proversity_reports/api/v0/views.py]: https://github.com/edx/opaque-keys/blob/master/opaque_keys/__init__.py#L167
course-v1:APP+APP001.3+2019

File: openedx_proversity_reports/api/v0/views.py [hunk 11]
```diff
@@ -242,3 +247,79 @@ def post(self, request):
...
+        get_jwt_authentication(),
+    )
+    permission_classes = (permissions.IsAuthenticated, get_staff_or_owner())
+
+    def post(self, request):
+        """
+        Return a JSON representation of the activity completion data by user.
+
+        **Params**
+            users: List of the email of the users. This parameter must contain at least one value.
```
[the remark this answers is not in the document]
Comment (Squirrel18) [reply openedx_proversity_reports/api/v0/views.py]: I mean, it needs at least one value to get any data. Changed.

File: openedx_proversity_reports/serializers.py (hunk 2 above)
Comment (Squirrel18) [reply openedx_proversity_reports/serializers.py]: Changed.

File: openedx_proversity_reports/utils.py (hunk 3 above)
Comment (Squirrel18) [reply openedx_proversity_reports/utils.py]: Changed.

File: openedx_proversity_reports/utils.py (hunk 5 above)
Comment (Squirrel18) [reply openedx_proversity_reports/utils.py]: To get the course enrollments only for students.

File: openedx_proversity_reports/utils.py (hunk 6 above)
Comment (Squirrel18) [reply openedx_proversity_reports/utils.py]: Nice catch, I didn't notice, addressed.

**Comment by Squirrel18** [comment]
PR addresed. @andrey-canon
